pkg/common-controller/groupsnapshot_controller_helper.go

```diff
@@ -485,6 +493,204 @@ func (ctrl *csiSnapshotCommonController) syncUnreadyGroupSnapshot(groupSnapshot
 	return nil
 }
 
+func (ctrl *csiSnapshotCommonController) createSnapshotsForGroupSnapshotContent(
+	groupSnapshotContent *crdv1alpha1.VolumeGroupSnapshotContent,
+	groupSnapshot *crdv1alpha1.VolumeGroupSnapshot,
+) (*crdv1alpha1.VolumeGroupSnapshotContent, error) {
+	type snapshotContentData struct {
+		snapshotContentName string
+		persistentVolume    *v1.PersistentVolume
+	}
+
+	volumeHandleData := make(map[string]snapshotContentData)
+
+	// No status is present. We need to wait for the snapshotter sidecar
+	// to fill it.
+	if groupSnapshotContent.Status == nil {
+		return groupSnapshotContent, nil
+	}
+
+	// If we the back-references are set, we already created the
+	// required objects. No need to do it another time.
+	if len(groupSnapshotContent.Status.PVVolumeSnapshotContentList) > 0 {
+		return groupSnapshotContent, nil
+	}
+
+	// The contents of the volume group snapshot class are needed to set the
+	// metadata containing the secrets to recover the snapshots
+	if groupSnapshot.Spec.VolumeGroupSnapshotClassName == nil {
+		return groupSnapshotContent, fmt.Errorf(
+			"createSnapshotsForGroupSnapshotContent: internal error: cannot find reference to volume group snapshot class")
+	}
+
+	groupSnapshotClass, err := ctrl.groupSnapshotClassLister.Get(*groupSnapshot.Spec.VolumeGroupSnapshotClassName)
+	if err != nil {
+		return groupSnapshotContent, fmt.Errorf(
+			"createSnapshotsForGroupSnapshotContent: failed to get volume snapshot class %s: %q",
+			*groupSnapshot.Spec.VolumeGroupSnapshotClassName, err)
+	}
+
+	groupSnapshotSecret, err := utils.GetGroupSnapshotSecretReference(
+		utils.GroupSnapshotterSecretParams,
+		groupSnapshotClass.Parameters,
+		groupSnapshotContent.GetObjectMeta().GetName(), nil)
+	if err != nil {
+		return groupSnapshotContent, fmt.Errorf(
+			"createSnapshotsForGroupSnapshotContent: failed to get secret reference for group snapshot content %s: %v",
+			groupSnapshotContent.Name, err)
+	}
+
+	pvs, err := ctrl.client.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return groupSnapshotContent, fmt.Errorf("createSnapshotsForGroupSnapshotContent: error get PersistentVolumes list from API server: %v", err)
+	}
+
+	// Phase 1: create the VolumeSnapshotContent and VolumeSnapshot objects
+	klog.V(4).Infof(
+		"createSnapshotsForGroupSnapshotContent[%s]: creating volumesnapshots and volumesnapshotcontent for group snapshot content",
+		groupSnapshotContent.Name)
+
+	for _, snapshot := range groupSnapshotContent.Status.VolumeSnapshotHandlePairList {
+		snapshotHandle := snapshot.SnapshotHandle
+		volumeHandle := snapshot.VolumeHandle
+
+		pv := utils.GetPersistentVolumeFromHandle(pvs, groupSnapshotContent.Spec.Driver, volumeHandle)
+		if pv == nil {
+			klog.Errorf(
```

Comment:
This is an interesting case. A snapshot exists, but its PV has been deleted. How useful is to have a  VolumeSnapshot  / VolumeSnapshotContent for it, if the user cannot possibly guess from which PVC the snapshot is?

My gut feeling would be to record also the source PVC + PV name in `VolumeSnapshotHandlePairList` to be able to reconstruct the VolumeSnapshotContent / VolumeSnapshot with a link to the PV / PVC

This should be discussed in the KEP to keep track of the decision.

Reply:
We retrieve PV here to populate groupSnapshotContent.Status.PVVolumeSnapshotContentList and set volumeSnapshotContent.Spec.SourceVolumeMode.
- The SourceVolumeMode will have an impact on applications that rely on changing volume mode when creating PVC from VolumeSnapshot. Without it being set, it falls back to the default behavior (for backward compatibility) which is not to check it. 
- Regarding PVVolumeSnapshotContentList, we have an action item to see if we can get rid of it after making all other changes.
I'm not sure if it is worth it to add source PVC and PV name in VolumeSnapshotHandlePairList.  That sounds like the wrong place to store this info. We can discuss more about it.

Reply:
Let me open an issue and we'll get back to this one later.